File: GFG/14th Dec- Painting the Fence/PaintingtheFence.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

// } Driver Code Ends
//User function Template for Java

class Solution {
    long countWays(int n, int k) {
        long total = k, same = 0, diff = k, mod = (long) 1e9 + 7;
        for (int i = 2; i <= n; i++) {
            same = diff;
            diff = (total % mod * (k - 1) % mod) % mod;
            total = (same + diff) % mod;
        }
        return total;
    }
}

// { Driver Code Starts.

// Driver class
class PaintingtheFence {

    // Driver code
    public static void main(String[] args) throws IOException {
        // Taking input using buffered reader
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int testcases = Integer.parseInt(br.readLine());
        // looping through all testcases
        while (testcases-- > 0) {
            // int n=Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] a2 = line.trim().split("\\s+");
            int n = Integer.parseInt(a2[0]);
            int k = Integer.parseInt(a2[1]);

            Solution ob = new Solution();
            // ArrayList<Long> ans=ob.smallestDifferenceTriplet(a,b,c,n);
            long ans = ob.countWays(n, k);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends

File: GFG/14th Dec- Painting the Fence/PaintingtheFence.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
#include<stdio.h>
#include<math.h>
using namespace std;

// } Driver Code Ends
class Solution{
    public:
    long long countWays(int n, int k){
        if(n == 1)
            return k;
            
        vector<vector<long long>> dp(n + 1, vector<long long> (2, 1));
        const long long mod = 1e9 + 7;
        
        for(int i = n - 1; i > 0; i--){
            dp[i][0] = (((k - 1) * dp[i + 1][0]) % mod + dp[i + 1][1]) % mod;
            
            dp[i][1] = ((k - 1) * dp[i + 1][0]) % mod;
        }
        
        return dp[0][0] = (k * dp[1][0]) % mod;
    }
};


//{ Driver Code Starts.

int main()
{
	
	int t;
	cin>>t;
	while(t--)
	{
		int n,k;
		cin>>n>>k;
		Solution ob;
		cout<<ob.countWays(n,k)<<endl;
	}
	return 0;
}
// } Driver Code Ends